sql/alps_submitter.go

```diff
@@ -329,30 +383,191 @@ func generateEstimatorCreator(estimator string, attrs []*attribute) (string, err
 	return fmt.Sprintf("tf.estimator.%s(%s)", estimator, strings.Join(cl, ",")), nil
 }
 
-func genALPS(w io.Writer, pr *extendedSelect, fts fieldTypes, db *DB) error {
-	// TODO(uuleon): codegen alps
+func newALPSTrainFiller(pr *extendedSelect) (*alpsFiller, error) {
+	fcList, fsMap, err := resolveTrainColumns(&pr.columns)
+	if err != nil {
+		return nil, err
+	}
+
+	fssCode := make([]string, 0, len(fsMap))
+	for _, fs := range fsMap {
+		fssCode = append(fssCode, fs.ToString())
+	}
+
+	fcCode, err := generateFeatureColumnCode(fcList)
+	if err != nil {
+		return nil, err
+	}
+
+	attrs, err := resolveTrainAttribute(&pr.attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	odpsAttrs := filter(attrs, "odps")
+	odpsMap := make(map[string]string, len(odpsAttrs))
+	for _, a := range odpsAttrs {
+		odpsMap[a.Name] = a.Value.(string)
+	}
+
+	trainSpecAttrs := filter(attrs, "train_spec")
+	trainMap := make(map[string]string, len(trainSpecAttrs))
+	for _, a := range trainSpecAttrs {
+		trainMap[a.Name] = a.Value.(string)
+	}
+
+	datasetAttrs := filter(attrs, "dataset")
+	datasetMap := make(map[string]string, len(datasetAttrs))
+	for _, a := range datasetAttrs {
+		datasetMap[a.Name] = a.Value.(string)
+	}
+
+	estimatorAttrs := filter(attrs, "estimator")
+
+	estimatorCode, err := generateEstimatorCreator(pr.estimator, estimatorAttrs)
+	if err != nil {
+		return nil, err
+	}
+
+	estimatorCode = strings.Replace(estimatorCode, "\"FC\"", "feature_columns", 1)
+
+	tableName := pr.tables[0]
+
+	fields := make([]string, len(pr.fields))
+	for idx, f := range pr.fields {
+		fields[idx] = fmt.Sprintf("\"%s\"", f)
+	}
+
+	y := &featureSpec{
+		FeatureName: pr.label,
+		IsSparse:    false,
+		Shape:       []int{1},
+		DType:       "int",
+		Delimiter:   ","}
+
+	dir, err := ioutil.TempDir("/tmp", "alps_scratch_dir")
```

Comment:
It would great to remove the temporary directory after the training. Maybe create a temporary directory in `trainALPS` so that we can remove it using defer.
```go
dir, err := ioutil.TempDir("/tmp", "alps_scratch_dir")
if e != nil {
	return e
}
defer os.RemoveAll(dir)
```

Reply:
Also, is it possible to combine `dir` and `cwd`?

Reply:
The `scratch_dir` will be deleted after model saving in the next PR.

Reply:
> Also, is it possible to combine `dir` and `cwd`?

In my opinion, no matter which `submitter` is used, the "current working directory" which is stateful should be managed (created, saved, deleted) by `runExtendedSQL` in SQLFlow, but the `scratch_dir` which is an inner attribute in `ALPS submitter`.